docs/client-specifications/error-handling.md

```diff
@@ -1,3 +1,143 @@
+  BAD_REQUEST_ERROR = 'BAD_REQUEST_ERROR',                 // message: `The request was invalid; please contact Momento: ${detail}`
+  TIMEOUT_ERROR = 'TIMEOUT_ERROR',                         // message: `The client's configured timeout was exceeded; you may need to use a Configuration with more lenient timeouts.  Timeout value: ${timeout}`
+  SERVER_UNAVAILABLE = 'SERVER_UNAVAILABLE',               // message: `The server was unable to handle the request; consider retrying.  If the error persists, please contact Momento.`
+  CLIENT_RESOURCE_EXHAUSTED = 'CLIENT_RESOURCE_EXHAUSTED', // message: `A client resource (most likely memory) was exhausted.  If you are executing a high volume of concurrent requests or using very large object sizes, your Configuration may need to be updated to allocate more memory.  Please contact Momento for assistance.`
+}
+```
+
+The `MomentoErrorTransportDetails` interface should capture the relevant gRPC information, for Momento debugging purposes:
+
+```typescript
+interface MomentoErrorTransportDetails {
+  grpc: MomentoGrpcErrorDetails
+}
+
+interface MomentoGrpcErrorDetails {
+  code: grpc.Status;
+  details: string;
+  metadata: grpc.Metadata;
+}
+```
+
+# Exposing Errors In Return Values
+
+**In languages that support sealed / algebraic data types, these should always be preferred.  In this document we are
+using TypeScript for our examples, so this illustrates more of a worst-case scenario for languages that lack such features.**
+
+`SimpleCacheClient` functions should return custom data types that are defined in the SDKs - they should not directly
+return generated gRPC response objects.
+
+## `CreateCache` and similar operations
+
+The `CreateCache` return value should have a `status` field, whose value is an enumeration including `SUCCESS` and `ERROR`.
+If the status is `ERROR`, then a `.error` field of type `Error` should be available:
+
+```typescript
+const result = client.createCache("myCache");
+if (result.status === CreateCacheStatus.SUCCESS) {
+  console.log(`Cache created`);
+} else if (result.status === CreateCacheStatus.ERROR) {
+  const error = result.error;
+  console.log(`An error occurred while attempting to create cache 'myCache': ${error.message}`);
+}
+```
+
+## `Get` and similar operations
+
+The `Get` return value should have a `status` field, whose value is an enumeration including `HIT`, `MISS` and `ERROR`.
+If the status is `ERROR`, then a `.error` field of type `Error` should be available:
+
+```typescript
+const result = client.get('myCache', 'myKey');
+if (result.status === CacheGetStatus.HIT) {
+  console.log(`Cache hit!  value: ${result.string()}`);
+} else if (result.status === CacheGetStatus.MISS) {
+  console.log(`Cache miss!  Time to populate the cache!`);
+} else if (result.status === CacheGetStatus.ERROR) {
```

Comment:
Just for my understanding, are you intending that we commit for eternity to

There are 3 possible Status returns for get: HIT, MISS and ERROR. No more, no less.
There are 2 possible Status returns for CreateCache: SUCCESS and ERROR. No more, no less.

If so, that's cool & I really like this. If not we might want to think harder =)

Reply:
I wouldn't go so far as to say "commit for eternity".  I am not aware of anything right now that would require us to add more values to those enums so I would not expect them to expand in the future, but I'm not totally clear on what you mean when you say "we might want to think harder", w/rt the unlikely possibility that we might need to add more in the future?

Reply:
Your example code does not have a case for an unhandled Status. If someone were to use this pattern here and we added a Status then their code would silently not do its job.

If we're not going to commit to it for eternity, then to account for the unlikely future & leaning on the new tenet above you can consider something like:
```suggestion
} else {
  console.log(`An error occurred while attempting to get key 'myKey' from cache 'myCache': ${error}`);
}
```

Reply:
i hear you.  i can add an else.  but keep in mind that the target audience for this doc is SDK authors.  I'm trying to focus on what things they need to make sure they include in their SDK implementations.  I am not trying to use this as example docs for end users.  In the former case, there are details about the return values that are important to make visible so that SDK authors make sure to implement things properly, and those are more visible in the code samples if I don't lump a bunch of stuff into a generic `else` branch.

End-user docs and examples will be provided for each language, closer to where their SDKs are defined.  And they can use the more strict compile-time branch checking where available.